File: ardoc/ardoc_kill_fam.pl
#!/usr/bin/env perl
use Env;
use Cwd;
use subs qw/get_pids/;

sub get_pids {

    my($procs, @kids) = @_;

    my @pids;
    foreach my $kid (@kids) {
       foreach my $proc (@$procs) {
           if ($proc->{'ppid'} == $kid && $proc->{'pgid'} == $group_id ) {
#	   if ($proc->{'ppid'} == $kid ) {
               my $pid = $proc->{'pid'};
#               print " FFFF $pid \n";
               push @pids, $pid, get_pids $procs, $pid;
           }
       }
    }
    @pids;
}


if ($#ARGV < 1) {
    print "ardoc_kill_fam: arguments required: process number(s), signal \n";
    exit 1;}

my $signal = $ARGV[$#ARGV];
$fields = $#ARGV - 1; 

if ( $signal < 0 || $signal > 15 ) { 
   print "ardoc_kill_fam: signal (last argument) must be between 0 and 15 \n";
   exit 1;}

for ($mmm=0; $mmm <= $fields; $mmm++){

my $procnum = $ARGV[$mmm];
print "ardoc_kill_fam: handling process $procnum ... \n";
my $UID = $>;
$group_id;
my @pids;
my @procs1=();
push @pids, $procnum;

@output=`ps -ww -u $UID  -o pid,ppid,pgid,args`;

#foreach $proc (@output)
#{print "AAA $proc \n";} 

foreach $line (@output)
{
    @fields = split(" ", $line);
    $apid = $fields[0];
    next if ( $apid eq "PID");
    $appid = $fields[1];
    $apgid = $fields[2]; 
    if ( $apid == $procnum ){$group_id = $apgid;}
    @aarg=();
	for ($m=3; $m <= $#fields; $m++){
	    push(@aarg, $fields[$m]);
	}
    $entry1={};
    $entry1->{'pid'}=$apid;
    $entry1->{'ppid'}=$appid;
    $entry1->{'pgid'}=$apgid;
    $entry1->{'args'}=@aarg;
    @kd=();  
          foreach $line1 (@output)
          {
              @fields = split(" ", $line1);
	      $apidk = $fields[0];
              $appidk = $fields[1];
	      next if ( $appidk ne $apid );
	      push(@kd, $apidk);
           }  
    $entry{'kids'}=[ @kd ];
    $entry1->{'kids'}=[ @kd ];
#    print "SDD $procnum $apid $appid $apgid $entry1->{'pid'} \n " ;
#    print "KDD @kd @{$entry1->{'kids'}}\n";
    push(@procs1,$entry1); 
}

#foreach $proc1 (@procs1)
#{$a=$proc1{'pid'};
#print "OOOO $proc1->{'pid'} \n";
# @df=@{$proc1->{'kids'}};
#print "OOAA @df \n";
#}
#print "GID $group_id\n";

my(@procs)=@procs1;
my(@kids) = get_pids \@procs, @pids;

@pids = (@pids, @kids);
foreach $k (@pids)
{print "ardoc_kill_fam: killing $k with signal $signal \n";
#system("kill -9 $k"); 
}
#print "IDS $) xxx $> xxx $<\n";

kill $signal, @pids;
}

File: ardoc/ardoc_kill_fam.py
#!/usr/bin/env python3
import os
import sys
import subprocess
import signal

def get_pids(procs, pids_to_check, group_id):
    """
    Recursively find all descendant processes for a given list of PIDs
    within the same process group.
    """
    all_descendants = []
    for pid_to_check in pids_to_check:
        children = []
        for proc in procs:
            if proc['ppid'] == pid_to_check and proc['pgid'] == group_id:
                children.append(proc['pid'])
        
        if children:
            all_descendants.extend(children)
            all_descendants.extend(get_pids(procs, children, group_id))
            
    return all_descendants

def main():
    if len(sys.argv) < 3:
        print("ardoc_kill_fam.py: arguments required: process number(s), signal")
        sys.exit(1)

    try:
        sig = int(sys.argv[-1])
        if not (0 <= sig <= 15):
             raise ValueError
    except ValueError:
        print("ardoc_kill_fam.py: signal (last argument) must be an integer between 0 and 15")
        sys.exit(1)

    proc_numbers_to_kill = []
    for arg in sys.argv[1:-1]:
        try:
            proc_numbers_to_kill.append(int(arg))
        except ValueError:
            print(f"ardoc_kill_fam.py: invalid process number: {arg}")
            continue

    uid = os.getuid()
    
    for procnum in proc_numbers_to_kill:
        print(f"ardoc_kill_fam.py: handling process {procnum} ...")
        
        try:
            output = subprocess.check_output(['ps', '-ww', '-u', str(uid), '-o', 'pid,ppid,pgid,args'], text=True)
        except subprocess.CalledProcessError as e:
            print(f"Error running ps: {e}")
            continue

        procs = []
        lines = output.strip().split('\n')
        header = lines[0].strip().split()
        
        group_id = None

        for line in lines[1:]:
            parts = line.strip().split(None, 3)
            if len(parts) < 3:
                continue
            
            pid = int(parts[0])
            ppid = int(parts[1])
            pgid = int(parts[2])
            
            if pid == procnum:
                group_id = pgid

            procs.append({'pid': pid, 'ppid': ppid, 'pgid': pgid})

        if group_id is None:
            print(f"ardoc_kill_fam.py: could not find process {procnum} for user {uid}")
            continue

        pids_to_kill = [procnum]
        descendants = get_pids(procs, pids_to_kill, group_id)
        pids_to_kill.extend(descendants)

        for pid in set(pids_to_kill):
            print(f"ardoc_kill_fam.py: killing {pid} with signal {sig}")
            try:
                os.kill(pid, sig)
            except ProcessLookupError:
                print(f"ardoc_kill_fam.py: process {pid} not found.")
            except PermissionError:
                print(f"ardoc_kill_fam.py: permission denied to kill process {pid}.")
            except Exception as e:
                print(f"ardoc_kill_fam.py: failed to kill process {pid}: {e}")


if __name__ == "__main__":
    main()
